sdk/src/packet.rs

```diff
@@ -38,111 +42,143 @@ pub struct Meta {
-    buffer: [u8; PACKET_DATA_SIZE],
-    pub meta: Meta,
-}
+pub trait BasePacket: Default + Clone + Sync + Send + Eq + fmt::Debug {
+    const DATA_SIZE: usize;
+    fn populate_packet<T: Serialize>(&mut self, dest: Option<&SocketAddr>, data: &T) -> Result<()>;
 
-impl Packet {
-    pub fn new(buffer: [u8; PACKET_DATA_SIZE], meta: Meta) -> Self {
-        Self { buffer, meta }
-    }
+    fn meta(&self) -> &Meta;
+    fn meta_mut(&mut self) -> &mut Meta;
+    fn buffer(&self) -> &[u8];
+
+    /// Returns a mutable reference to the entirety of the underlying buffer to
+    /// write into. The caller is responsible for updating Packet.meta.size
+    /// after writing to the buffer.
+    fn buffer_mut(&mut self) -> &mut [u8];
 
     /// Returns an immutable reference to the underlying buffer up to
     /// packet.meta.size. The rest of the buffer is not valid to read from.
     /// packet.data(..) returns packet.buffer.get(..packet.meta.size).
     /// Returns None if the index is invalid or if the packet is already marked
     /// as discard.
     #[inline]
-    pub fn data<I>(&self, index: I) -> Option<&<I as SliceIndex<[u8]>>::Output>
+    fn data<I>(&self, index: I) -> Option<&<I as SliceIndex<[u8]>>::Output>
     where
         I: SliceIndex<[u8]>,
     {
         // If the packet is marked as discard, it is either invalid or
         // otherwise should be ignored, and so the payload should not be read
         // from.
-        if self.meta.discard() {
+        if self.meta().discard() {
             None
         } else {
-            self.buffer.get(..self.meta.size)?.get(index)
+            self.buffer().get(..self.meta().size)?.get(index)
         }
     }
 
-    /// Returns a mutable reference to the entirety of the underlying buffer to
-    /// write into. The caller is responsible for updating Packet.meta.size
-    /// after writing to the buffer.
-    #[inline]
-    pub fn buffer_mut(&mut self) -> &mut [u8] {
-        debug_assert!(!self.meta.discard());
-        &mut self.buffer[..]
-    }
-
-    pub fn from_data<T: Serialize>(dest: Option<&SocketAddr>, data: T) -> Result<Self> {
-        let mut packet = Packet::default();
+    fn from_data<T: Serialize>(dest: Option<&SocketAddr>, data: T) -> Result<Self> {
+        let mut packet = Self::default();
         Self::populate_packet(&mut packet, dest, &data)?;
         Ok(packet)
     }
 
-    pub fn populate_packet<T: Serialize>(
-        &mut self,
-        dest: Option<&SocketAddr>,
-        data: &T,
-    ) -> Result<()> {
-        debug_assert!(!self.meta.discard());
-        let mut wr = io::Cursor::new(self.buffer_mut());
-        bincode::serialize_into(&mut wr, data)?;
-        self.meta.size = wr.position() as usize;
-        if let Some(dest) = dest {
-            self.meta.set_socket_addr(dest);
-        }
-        Ok(())
-    }
-
-    pub fn deserialize_slice<T, I>(&self, index: I) -> Result<T>
+    fn deserialize_slice<T, I>(&self, index: I) -> Result<T>
```

Comment:
Can keep the functions in the same order as original to reduce changes and easier to compare. Like move populate_packet here

Reply:
nvm, saw you moved it into a macro.